paddle/fluid/platform/nccl_helper.h

```diff
@@ -160,6 +160,151 @@ struct NCCLContextMap {
+                    const std::vector<ncclUniqueId *> &nccl_ids,
+                    size_t trainers_num, size_t trainer_id) {
+    if (nccl_ids.size() == 0) {
+      auto ptr = new platform::NCCLContextMap(places);
+      VLOG(1) << "init local trainer";
+      flat_ctxs_.push_back(ptr);
+      return;
+    }
+
+    for (size_t i = 0; i < nccl_ids.size(); i++) {
+      auto ptr = new platform::NCCLContextMap(places, nccl_ids[i], trainers_num,
+                                              trainer_id);
+      VLOG(1) << "init trainer_id:" << trainer_id << ", comm no:" << i;
+      flat_ctxs_.push_back(ptr);
+    }
+  }
+
+  void InitHierarchicalCtxs(const std::vector<platform::Place> &places,
+                            ncclUniqueId *inter_nccl_id,
+                            const std::vector<ncclUniqueId *> &exter_nccl_id,
+                            size_t trainers_num, size_t trainer_id,
+                            size_t inter_trainers_num,
+                            size_t exter_trainers_num) {
+    PADDLE_ENFORCE(trainers_num == inter_trainers_num * exter_trainers_num,
+                   "trainers_num:%llu != inter_trainers_num:%llu * "
+                   "exter_trainers_num:%llu",
+                   trainers_num, inter_trainers_num, exter_trainers_num);
+
+    PADDLE_ENFORCE(inter_trainers_num > 1, "inter_trainers_num:%llu must > 1",
+                   inter_trainers_num);
+
+    int inter_trainer_id = trainer_id % inter_trainers_num;
+    VLOG(1) << "init inter_trainer_id:" << inter_trainer_id;
+    auto local = new NCCLContextMap(places, inter_nccl_id, inter_trainers_num,
+                                    inter_trainer_id);
+
+    h_inter_ctxs_.push_back(local);
+
+    int exter_trainer_id = -1;
+    if (trainer_id % inter_trainers_num == 0) {
+      exter_trainer_id = trainer_id / inter_trainers_num;
+    }
+
+    if (exter_trainer_id >= 0) {
+      for (size_t i = 0; i < exter_nccl_id.size(); i++) {
+        auto ex = new NCCLContextMap(places, exter_nccl_id[i],
+                                     exter_trainers_num, exter_trainer_id);
+        VLOG(1) << "init exter_trainer_id:" << exter_trainer_id
+                << ", comm no:" << i;
+        h_exter_ctxs_.push_back(ex);
+      }
+    }
+  }
+
+  bool NeedExterAllReduce() const { return h_exter_ctxs_.size() > 0; }
+
+  NCCLContextMap *GetHierarchicalInterCtx(size_t run_order) const {
+    return h_inter_ctxs_[run_order % h_inter_ctxs_.size()];
+  }
+
+  NCCLContextMap *GetHierarchicalExterCtx(size_t run_order) const {
+    return h_exter_ctxs_[run_order % h_exter_ctxs_.size()];
+  }
+
+  std::vector<NCCLContextMap *> *GetHierarchicalInterCtxs() {
+    return &h_inter_ctxs_;
+  }
+
+  std::vector<NCCLContextMap *> *GetHierarchicalExterCtxs() {
+    return &h_exter_ctxs_;
+  }
+
+ protected:
+  std::vector<NCCLContextMap *> flat_ctxs_;
+
+  std::vector<NCCLContextMap *> h_inter_ctxs_;
+  std::vector<NCCLContextMap *> h_exter_ctxs_;
```

Comment:
Please explain the difference of those NCCLContextMaps.

Reply:
Done.Thanks!